crates/kv-store/src/sstable.rs

```diff
@@ -0,0 +1,1026 @@
+                    if !iter.has_next_back() {
+                        iter.next();
+                    }
+                    (end_idx, None, Some(end))
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_back_to_key(block, end);
+                    (end_idx, Some(iter), Some(end))
+                }
+            }
+            Bound::Unbounded => {
+                let end_idx = table.meta.len() - 1;
+                if end_idx == table_idx {
+                    (end_idx, None, None)
+                } else {
+                    let block = table.read_block_cached(end_idx);
+                    let iter = BlockIter::new_seek_to_first(block);
+                    (end_idx, Some(iter), None)
+                }
+            }
+        };
+
+        let mut ans = if let Some(end_iter) = end_iter {
+            debug_assert!(end_idx > table_idx);
+            SsTableIter {
+                table,
+                next_block_iter: iter,
+                next_block_idx: table_idx,
+                back_block_iter: end_iter,
+                back_block_idx: end_idx as isize,
+                next_first: false,
+            }
+        } else {
+            debug_assert!(end_idx == table_idx);
+            SsTableIter {
+                table,
+                next_block_iter: iter.clone(),
+                next_block_idx: table_idx,
+                back_block_iter: iter,
+                back_block_idx: end_idx as isize,
+                next_first: true,
+            }
+        };
+        // the current iter is empty, but has next iter. we need to skip the empty iter
+        ans.skip_next_empty();
+        ans.skip_next_back_empty();
+
+        if let Some(key) = excluded {
+            if ans.has_next() && ans.next_key().unwrap() == key {
+                ans.next();
+            }
+        }
+        if let Some(key) = end_excluded {
+            if ans.has_next_back() && ans.next_back_key().unwrap() == key {
+                ans.next_back();
+            }
+        }
+
+        // need to skip empty block
+        if ans.has_next() && !ans.next_block_iter.has_next() {
+            ans.next();
+        }
+
+        if ans.has_next_back() && !ans.next_first && !ans.back_block_iter.has_next_back() {
+            ans.next_back();
+        }
+        ans
+    }
+
+    fn skip_next_empty(&mut self) {
+        while self.has_next() && !self.next_block_iter.has_next() {
+            self.next();
+        }
+    }
+
+    fn skip_next_back_empty(&mut self) {
+        while self.has_next_back() && !self.next_first && !self.back_block_iter.has_next_back() {
+            self.next_back();
+        }
+    }
+
+    fn has_next(&self) -> bool {
+        self.next_block_iter.has_next() || (self.next_block_idx as isize) < self.back_block_idx
+    }
+
+    pub fn next_key(&self) -> Option<Bytes> {
```

Comment:
I think it's probably better to call these methods as `peek_next_key()`.

Otherwise, it's hard to tell which one actually modify the iterator when seeing a bunch of `next_key()` `next_back()` `next_back_key()`